File: src/lib/api/worktree.ts
import { getTransport } from '$lib/api';
import type {
  WorktreeInfo,
  WorktreeCreateResult,
  WorktreeRemoveResult,
  WorktreePruneResult,
} from '$lib/api/types';

export async function listWorktrees(path: string): Promise<WorktreeInfo[]> {
  return getTransport().invoke('list_worktrees', { path });
}

export async function addWorktree(
  path: string,
  worktreePath: string,
  branch: string | null,
  newBranch: boolean,
): Promise<WorktreeCreateResult> {
  return getTransport().invoke('add_worktree', {
    path,
    worktreePath,
    branch,
    newBranch,
  });
}

export async function removeWorktree(
  path: string,
  worktreePath: string,
  force: boolean,
): Promise<WorktreeRemoveResult> {
  return getTransport().invoke('remove_worktree', {
    path,
    worktreePath,
    force,
  });
}

export async function lockWorktree(
  path: string,
  worktreePath: string,
  reason?: string,
): Promise<WorktreeInfo[]> {
  return getTransport().invoke('lock_worktree', {
    path,
    worktreePath,
    reason: reason ?? null,
  });
}

export async function unlockWorktree(
  path: string,
  worktreePath: string,
): Promise<WorktreeInfo[]> {
  return getTransport().invoke('unlock_worktree', {
    path,
    worktreePath,
  });
}

export async function pruneWorktrees(
  path: string,
  dryRun: boolean,
): Promise<WorktreePruneResult> {
  return getTransport().invoke('prune_worktrees', {
    path,
    dryRun,
  });
}

File: crates/gitron-server/src/routes/worktree.rs
use axum::http::StatusCode;
use axum::Json;
use serde::Deserialize;

use gitron_core::git::{worktree, types::*};

#[derive(Deserialize)]
pub struct ListRequest {
    path: String,
}

#[derive(Deserialize)]
pub struct AddRequest {
    path: String,
    #[serde(rename = "worktreePath")]
    worktree_path: String,
    branch: Option<String>,
    #[serde(rename = "newBranch")]
    new_branch: bool,
}

#[derive(Deserialize)]
pub struct RemoveRequest {
    path: String,
    #[serde(rename = "worktreePath")]
    worktree_path: String,
    force: bool,
}

#[derive(Deserialize)]
pub struct LockRequest {
    path: String,
    #[serde(rename = "worktreePath")]
    worktree_path: String,
    reason: Option<String>,
}

#[derive(Deserialize)]
pub struct UnlockRequest {
    path: String,
    #[serde(rename = "worktreePath")]
    worktree_path: String,
}

#[derive(Deserialize)]
pub struct PruneRequest {
    path: String,
    #[serde(rename = "dryRun")]
    dry_run: bool,
}

pub async fn list_worktrees(
    Json(req): Json<ListRequest>,
) -> Result<Json<Vec<WorktreeInfo>>, (StatusCode, String)> {
    let result = worktree::list_worktrees(&req.path).map_err(err)?;
    Ok(Json(result))
}

pub async fn add_worktree(
    Json(req): Json<AddRequest>,
) -> Result<Json<WorktreeCreateResult>, (StatusCode, String)> {
    let result = worktree::add_worktree(
        &req.path,
        &req.worktree_path,
        req.branch.as_deref(),
        req.new_branch,
    )
    .map_err(err)?;
    Ok(Json(result))
}

pub async fn remove_worktree(
    Json(req): Json<RemoveRequest>,
) -> Result<Json<WorktreeRemoveResult>, (StatusCode, String)> {
    let result =
        worktree::remove_worktree(&req.path, &req.worktree_path, req.force).map_err(err)?;
    Ok(Json(result))
}

pub async fn lock_worktree(
    Json(req): Json<LockRequest>,
) -> Result<Json<Vec<WorktreeInfo>>, (StatusCode, String)> {
    worktree::lock_worktree(&req.path, &req.worktree_path, req.reason.as_deref())
        .map_err(err)?;
    let list = worktree::list_worktrees(&req.path).map_err(err)?;
    Ok(Json(list))
}

pub async fn unlock_worktree(
    Json(req): Json<UnlockRequest>,
) -> Result<Json<Vec<WorktreeInfo>>, (StatusCode, String)> {
    worktree::unlock_worktree(&req.path, &req.worktree_path).map_err(err)?;
    let list = worktree::list_worktrees(&req.path).map_err(err)?;
    Ok(Json(list))
}

pub async fn prune_worktrees(
    Json(req): Json<PruneRequest>,
) -> Result<Json<WorktreePruneResult>, (StatusCode, String)> {
    let result = worktree::prune_worktrees(&req.path, req.dry_run).map_err(err)?;
    Ok(Json(result))
}

fn err(e: impl std::fmt::Display) -> (StatusCode, String) {
    (StatusCode::INTERNAL_SERVER_ERROR, e.to_string())
}
